### PR TITLE
OTG Test - TE-4.1: Base Leader Election

### PR DESCRIPTION
**Pull request by octpetre**
Hello,

This is the OTG version of the # TE-4.1: Base Leader Election GRIBI test.

Thank you,

### CODE REVIEW

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go [hunk 1]
```diff
@@ -46,11 +49,14 @@ func TestMain(m *testing.M) {
...
-	ipv4PrefixLen = 30
-	instance      = "default"
-	ateDstNetCIDR = "198.51.100.0/24"
-	nhIndex       = 1
-	nhgIndex      = 42
+	ipv4PrefixLen   = 30
+	instance        = "default"
+	ateDstNetCIDR   = "198.51.100.0/24"
+	ateDstNetStart  = "198.51.100.0"
+	ateDstNetPrefix = 24
```
Comment (liulk) [inline feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:56]: Unused?

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go [hunk 2]
```diff
@@ -61,7 +67,8 @@ var (
 	}
 
 	atePort1 = attrs.Attributes{
-		Name:    "atePort1",
+		Name:    "port1",
+		MAC:     "02:00:01:01:01:01",
```
Comment (liulk) [inline feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:71]: Sorry, I forgot, is the MAC address optional? It would be better if they can be automatically assigned like the ATE API, since most tests don't need to know what MAC addresses to use.

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go [hunk 3]
```diff
@@ -61,7 +67,8 @@ var (
 	}
 
 	atePort1 = attrs.Attributes{
-		Name:    "atePort1",
+		Name:    "port1",
```
Comment (liulk) [inline feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:70]: Is atePort1 required to have a specific name in OTG?

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go [hunk 4]
```diff
@@ -85,14 +93,22 @@ var (
...
 
+var inputMap = map[attrs.Attributes]attrs.Attributes{
+	atePort1: dutPort1,
+	atePort2: dutPort2,
+	atePort3: dutPort3,
+}
+
 // configInterfaceDUT configures the interface with the Addrs.
 func configInterfaceDUT(i *telemetry.Interface, a *attrs.Attributes) *telemetry.Interface {
+	deviations.InterfaceEnabled = ygot.Bool(true)
```
Comment (liulk) [inline feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:111]: Let's revert this change because this is an Arista specific deviation.

When running the test against Arista, pass `--deviation_interface_enabled` as a flag.

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go [hunk 5]
```diff
@@ -128,55 +144,88 @@ func configureDUT(t *testing.T, dut *ondatra.DUTDevice) {
...
+	eth.SetPortName(dev.Name()).SetMac(atePort3.MAC)
+	ip = eth.Ipv4Addresses().Add().SetName(dev.Name() + ".IPv4")
+	ip.SetAddress(atePort3.IPv4).SetGateway(dutPort3.IPv4).SetPrefix(int32(atePort3.IPv4Len))
+
+	otg.PushConfig(t, top)
+	otg.StartProtocols(t)
 	return top
 }
 
+func waitOtgArpEntry(t *testing.T) {
```
Comment (liulk) [inline feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:178]: waitOtgArpEntry -> waitOTGARPEntry because of golang initialism.

https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go [hunk 6]
```diff
@@ -128,55 +144,88 @@ func configureDUT(t *testing.T, dut *ondatra.DUTDevice) {
...
+	// srcEndPoint is atePort1
+	otg := ate.OTG()
+	gwIp := inputMap[srcEndPoint].IPv4
+	waitOtgArpEntry(t)
+	dstMac := otg.Telemetry().Interface(srcEndPoint.Name + ".Eth").Ipv4Neighbor(gwIp).LinkLayerAddress().Get(t)
+	config.Flows().Clear().Items()
+	flowipv4 := config.Flows().Add().SetName("Flow")
+	flowipv4.Metrics().SetEnable(true)
+	flowipv4.TxRx().Port().SetTxName(srcEndPoint.Name).SetRxName(dstEndPoint.Name)
+	flowipv4.Size().SetFixed(512)
```
Comment (liulk) [inline feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:200]: Can we elide any of these settings like size and rate if they have a reasonable default?

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go [hunk 7]
```diff
@@ -211,25 +260,23 @@ func testIPv4LeaderActiveChange(ctx context.Context, t *testing.T, args *testArg
...
-	testTraffic(t, args.ate, args.top, srcEndPoint, dstEndPoint)
+	testTraffic(t, args.ate, args.top, atePort1, atePort3)
 
 	// Add an IPv4Entry for 198.51.100.0/24 pointing to ATE port-2 via gRIBI-A,
 	// ensure that the entry is ignored by the DUT.
 	t.Logf("Adding an IPv4Entry for %s pointing to ATE port-2 via gRIBI-A", ateDstNetCIDR)
-	args.clientA.AddNH(t, nhIndex+1, atePort2.IPv4, instance, fluent.ProgrammingFailed)
-	args.clientA.AddNHG(t, nhgIndex+1, map[uint64]uint64{nhIndex + 1: 1}, instance, fluent.ProgrammingFailed)
-	args.clientA.AddIPv4(t, ateDstNetCIDR, nhgIndex+1, instance, "", fluent.ProgrammingFailed)
+	args.clientA.AddNH(t, nhIndex, atePort2.IPv4, instance, fluent.ProgrammingFailed)
```
Comment (liulk) [inline feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:268]: Please port the index changes to OTG.

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go [hunk 8]
```diff
@@ -85,14 +93,22 @@ var (
...
 	atePort3 = attrs.Attributes{
-		Name:    "atePort3",
+		Name:    "port3",
+		MAC:     "02:00:03:01:01:01",
 		IPv4:    "192.0.2.10",
 		IPv4Len: ipv4PrefixLen,
 	}
 )
 
+var inputMap = map[attrs.Attributes]attrs.Attributes{
```
Comment (liulk) [inline feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:103]: inputMap -> gatewayMap?

**Review by liulk — CHANGES_REQUESTED** [review]

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go (hunk 1 above)
Comment (octpetre) [reply feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:56]: ateDstNetPrefix removed.

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go (hunk 2 above)
Comment (octpetre) [reply feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:71]: The MAC address is mandatory when creating OTG ethernet devices [OTG spec - set config](https://redocly.github.io/redoc/?url=https://raw.githubusercontent.com/open-traffic-generator/models/master/artifacts/openapi.yaml#tag/Configuration/operation/set_config)
Also I don't know why there are no defaults like in the ATE API but I'm asking to see if this could be implemented

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go (hunk 3 above)
Comment (octpetre) [reply feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:70]: No, but these names have to match the atedut_4.testbed file for ate ports ids, which are currently set to "port1", "port2, ... etc

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go (hunk 6 above)
Comment (octpetre) [reply feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:200]: Using default 64 for size, 1000 for pps and "continuous" for duration now

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go (hunk 7 above)
Comment (octpetre) [reply feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:268]: Done

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go (hunk 8 above)
Comment (octpetre) [reply feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go:103]: Done

File: feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go (hunk 4 above)
Comment (octpetre) [reply feature/gribi/otg_tests/base_leader_election_test/base_leader_election_test.go]: Done

**Review by liulk — APPROVED** [review]
